src/types/accessors/bits/semantics.jl

```diff
@@ -0,0 +1,148 @@
+
+$example
+
+Use [`$count`](@ref) to quickly determine the amount of $name in the
+model. See the documentation of [`$mapping`](@ref) for closer
+definition of the correspondence of $name and model variables.
+""",
+        ),
+        :(function $plural(a::AbstractMetabolicModel)::Vector{String}
+            x = collect(keys($mapping))
+            sort!(x)
+            x
+        end),
+    )
+
+    countfn = Expr(
+        :macrocall,
+        Symbol("@doc"),
+        source,
+        Expr(
+            :string,
+            :TYPEDSIGNATURES,
+            """
+
+Count of $name that the model describes, should be equal to the length of
+vector returned by [`$plural`].
+""",
+        ),
+        :(function $count(a::AbstractMetabolicModel)::Int
+            length($mapping)
+        end),
+    )
+
+    mappingfn = Expr(
+        :macrocall,
+        Symbol("@doc"),
+        source,
+        Expr(
+            :string,
+            :TYPEDSIGNATURES,
+            """
+
+Bipartite mapping of $name described by the model to the actual
+variables in the model. Returns a dictionary of $name assigned to the
+variable IDs and their linear coefficients. See the documentation of
+[`$plural`](@ref) for semantics.
+
+To improve the performance, you may want to use [`$mapping_mtx`](@ref).
+""",
+        ),
+        :(function $mapping(a::AbstractMetabolicModel)::Dict{String,Dict{String,Float64}}
+                Dict()
+            end),
+    )
+
+    mtxfn = Expr(
+        :macrocall,
+        Symbol("@doc"),
+        source,
+        Expr(
+            :string,
+            :TYPEDSIGNATURES,
+            """
+
+Bipartite mapping of $name described by the model to the actual
+variables in the model, described as a sparse matrix mapping with rows
+corresponding to model variables and columns corresponding to $name.
+
+By default, this is derived from [`$mapping`](@ref) in all models. For
+safety reasons, this is never automatically inherited by wrappers.
+""",
+        ),
+        :(function $mapping_mtx(a::AbstractMetabolicModel)::SparseMat
+            make_mapping_mtx(variables(a), $plural(a), $mapping(a))
+        end),
+    )
+
+    code = Expr(:block, pluralfn, countfn, mappingfn, mtxfn)
+
+    Base.eval(themodule, code)
+end
+
+macro make_variable_semantics(sym, name, example)
+    src = __source__
+    # TODO actually use the source
+    quote
+        $make_variable_semantics($Accessors, $src, $sym, $name, $example)
+    end
+end
+
+macro all_variables_are_reactions(mt)
+    m = esc(mt)
+    # TODO docs
+    quote
+        $Accessors.reactions(model::$m) = $Accessors.variables(model)
+        $Accessors.n_reactions(model::$m) = $Accessors.n_variables(model)
+        $Accessors.reaction_variables(model::$m) =
+            Dict(var => Dict(var => 1.0) for var in variables(model))
+        $Accessors.reaction_variables_matrix(model::$m) =
+            $SparseArrays.spdiagm(fill(1.0, $Accessors.n_variables(model)))
+    end
+end
```

Comment:
All of this needs a "how to for dummies" - I will test out if it is clear enough when I do the enzyme models later today

Reply:
I added some docstrings; dummies may find the basic examples with grep. There will also be docs.